doc/ifaces.1
.TH IFACES 1 "November 2023" "ifaces 1.x" "User Commands"
.SH NAME
ifaces \- List interface names
.SH SYNOPSIS
.B ifaces
[\fIOPTION\fR]
.SH DESCRIPTION
The \fBifaces(1)\fR tool prints out the list of interfaces available on your
computer.

.SH "COMMAND LINE OPTIONS"
.TP
\fB\-\-asterisk\fR
Print an asterisk (*) for the default address (instead of 0.0.0.0 or ::).

.TP
\fB\-d\fR, \fB\-\-default\fR
Only print information about the default interface.

.TP
\fB\-h\fR, \fB\-\-help\fR
Print a brief document about the tool usage, then exit.

.TP
\fB\-\-hide\-headers\fR
Do not print the headers of the output table.

.TP
\fB\-\-loopback\fR
Only print the name of the loopback interface (usually "lo").

.TP
\fB\-\-name-only\fR
Only print the interface names. No address, columns, headers.

.TP
\fB\-\-private\fR
Only print private interfaces.

.TP
\fB\-\-public\fR
Only print public interfaces.

.SH AUTHOR
Written by Alexis Wilke <alexis@m2osw.com>.
.SH "REPORTING BUGS"
Report bugs to <https://github.com/m2osw/libaddr/issues>.
.br
iplock home page: <https://snapwebsites.org/project/libaddr>.
.SH COPYRIGHT
Copyright \(co 2023-2025  Made to Order Software Corp.  All Rights Reserved
.br
License: GPLv3
.br
This is free software: you are free to change and redistribute it.
.br
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR ipv4-routes (1),
.BR validate-ip (1)


debian/changelog
libaddr (1.0.37.0~jammy) jammy; urgency=high

  * Fixed the tests and added a few more to slightly better coverage.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 02 Jun 2024 20:28:11 -0700

libaddr (1.0.36.0~jammy) jammy; urgency=high

  * Added the --version command line option to ifaces tool.
  * Added a manual page for the ifaces tool.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 18 Nov 2023 10:38:19 -0800

libaddr (1.0.35.0~jammy) jammy; urgency=high

  * Added tool to list interfaces with filtering support.
  * Added support for default IP address as "*".
  * Enhanced output of ip4_route.
  * Added a 5 minutes cache for the iface class.
  * Added missing #include <cstdint>
  * Added some #pragma -Wrestrict for lunar.
  * Bumped copyright to 2023.
  * Changed compat to latest (15).
  * Removed boost dependency.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 12 Nov 2022 17:43:55 -0800

libaddr (1.0.34.1~jammy) jammy; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 12 Nov 2022 17:43:55 -0800

libaddr (1.0.34.0~bionic) bionic; urgency=high

  * Added a set_group()/get_group() for Unix file-based sockets.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 05 Nov 2022 10:56:32 -0700

libaddr (1.0.33.0~bionic) bionic; urgency=high

  * Added a set_mode()/get_mode() for Unix file-based sockets.

 -- Alexis Wilke <alexis@m2osw.com>  Fri, 04 Nov 2022 21:48:18 -0700

libaddr (1.0.32.0~bionic) bionic; urgency=high

  * Added the to_cidr() to the addr_range class.
  * Added the optimize_vector() to remove equal/included addresses.
  * Renamed the unix.h as addr_unix.h so it works with gnu++17.
  * Added support of comments within the address.
  * Added support for comments introduced by ';'.
  * Added support for set_port() with a string.
  * Fixed the CIDR/mask parsing when the address was an IPv4 in an IPv6.
  * Output the mask "if needed" (if exactly 128 or 32).
  * Redesigned the ip to string code to use flags instead of an enum.
  * Added a sento() for UDP packets.
  * Added flags to know what was defined in the parser.
  * Added the .dirs files.
  * Various clean ups.

 -- Alexis Wilke <alexis@m2osw.com>  Mon, 24 Oct 2022 20:37:15 -0700

libaddr (1.0.31.1~bionic) bionic; urgency=high

  * For newer versions of g++, #include <limits>.

 -- Alexis Wilke <alexis@m2osw.com>  Mon, 11 Jul 2022 11:45:39 -0700

libaddr (1.0.31.0~bionic) bionic; urgency=high

  * Added support for newlines between addresses.
  * Added support for a '#' comment introducer (along the newline feature).
  * Added the to_string() on the addr_range class.
  * Added the to_string() for the addr_range::vector_t type.
  * Added support for address ranges in the parser (<addr>-<addr>).
  * Added ability to forbid address like masks (default now).
  * Correctly implemented the SORT_NO_EMPTY check.
  * Added support for ostream << operator including manip like parameters.
  * Added test to verify the input port as string.
  * Added error messages in the validator_address object.
  * Added cppthread as a dependency to support validator_address errors.
  * Redesign parsing of validator_address parameters.
  * Changed the set_protocol() to make use of getprotobyname_r().
  * Added a set_mask_count() to allow for a number instead of a 128 bit mask.
  * Allow for applying the reversed mask (i.e. addr |= ~mask).
  * Moved the network_type_t and string_ip_t outside of the addr class.
  * Fixed the range compare() function so it actually works.
  * Enhanced the parsing of IPv6 vs. IPv4 addresses.
  * Tested and fixed the sorting of IPs by the parser.
  * Added a range "is_defined()" if one of "from" or "to" is set.
  * Added a swap of the boundaries in addr_range.
  * Fixed union_if_possible() tests as "from" or "to" may be undefined.
  * Enhanced many tests and updated them to pass with the new features.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 11 Jun 2022 21:19:18 -0700

libaddr (1.0.30.0~bionic) bionic; urgency=high

  * Removed cppthread as a dependency. It's not required.

 -- Alexis Wilke <alexis@m2osw.com>  Fri, 20 May 2022 07:32:39 -0700

libaddr (1.0.29.0~bionic) bionic; urgency=high

  * Added a function to test an IP to know whether it's a broadcast IP.
  * Fixed documentation about the STRING_IP_... enumeration.
  * Added an advgetopt validator for IP addresses.
  * Made sure headers all use <...> for #include.
  * Support attaching an interface to an address.
  * Fixed compat to v10.
  * Cleaned up the cmake file.

 -- Alexis Wilke <alexis@m2osw.com>  Fri, 20 May 2022 07:15:47 -0700

libaddr (1.0.28.0~bionic) bionic; urgency=high

  * Added the Environment Variable Intro.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 05 Mar 2022 21:15:53 -0800

libaddr (1.0.27.0~bionic) bionic; urgency=high

  * Hold the original hostname in case it is rquired for SNI.
  * Added a get_family() function.
  * Renamed f_private_network_defined as just f_private_network.
  * Added a missing set_protocol() in the addr_parser.

 -- Alexis Wilke <alexis@m2osw.com>  Tue, 01 Mar 2022 21:02:10 -0800

libaddr (1.0.26.0~bionic) bionic; urgency=high

  * Properly clean up once done with tests.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 13 Feb 2022 14:12:09 -0800

libaddr (1.0.25.1~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 13 Feb 2022 13:54:16 -0800

libaddr (1.0.25.0~bionic) bionic; urgency=high

  * Implemented the sort so we can have IPv6 first, because we are expected
    to give priority to IPv6 addresses.
  * Added a sort flag to avoid having empty ranges returned.
  * Added a function to compute the size of a valid CIDR mask.
  * Added a function to convert the IP into an unsigned int128 value.
  * Allow the match() function to return true if this IP is ANY.
  * Added the is_next() and is_previous() functions.
  * Renamed the addr_parser::flag_t as allow_t.

tools/ifaces.cpp
// Network Address -- get routes and print them, similar to system `route`
// Copyright (c) 2012-2025  Made to Order Software Corp.  All Rights Reserved
//
// https://snapwebsites.org/project/libaddr
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

/** \file
 * \brief A tool to check the system interfaces.
 *
 * This tool is used to verify that our iface class works as expected.
 */

// libaddr
//
#include    <libaddr/iface.h>
#include    <libaddr/route.h>
#include    <libaddr/version.h>


// C++
//
#include    <iomanip>
#include    <iostream>
#include    <sstream>


// last include
//
#include    <snapdev/poison.h>



namespace
{

enum filter_t
{
    FILTER_ALL,             // no filtering
    FILTER_PUBLIC,          // only display interfaces with public IP addresses
    FILTER_PRIVATE,         // only display interfaces with private IP addresses
    FILTER_LOOPBACK,        // only display interfaces with the "lo" interface
    FILTER_DEFAULT,         // only display interfaces with the default route
};

filter_t g_filter = FILTER_ALL;
bool g_hide_headers = false;
bool g_asterisk = false;
bool g_name_only = false;

}

int main(int argc, char * argv[])
{
    for(int idx(1); idx < argc; ++idx)
    {
        if(strcmp(argv[idx], "-h") == 0
        || strcmp(argv[idx], "--help") == 0)
        {
            std::cout
                << "Usage: " << argv[0] << " [-opts]\n"
                   "where -opts is one or more of:\n"
                   "  --help | -h        print out this help screen.\n"
                   "  --default | -d     only print name of default interface.\n"
                   "  --hide-headers     do not print the headers.\n"
                   "  --public           only print name of public interfaces.\n"
                   "  --private          only print name of private interfaces.\n"
                   "  --loopback         only print name of loopback interface.\n"
                   "  --asterisk         print an asterisk for default addresses (instead of 0.0.0.0 or ::).\n"
                   "  --name-only        only print the name of the interface.\n"
                   "  --version | -V     print out the version of ifaces.\n"
                ;
            return 1;
        }
        else if(strcmp(argv[idx], "--hide-headers") == 0)
        {
            g_hide_headers = true;
        }
        else if(strcmp(argv[idx], "-d") == 0
             || strcmp(argv[idx], "--default") == 0)
        {
            g_filter = FILTER_DEFAULT;
        }
        else if(strcmp(argv[idx], "--public") == 0)
        {
            g_filter = FILTER_PUBLIC;
        }
        else if(strcmp(argv[idx], "--private") == 0)
        {
            g_filter = FILTER_PRIVATE;
        }
        else if(strcmp(argv[idx], "--loopback") == 0)
        {
            g_filter = FILTER_LOOPBACK;
        }
        else if(strcmp(argv[idx], "--asterisk") == 0)
        {
            g_asterisk = true;
        }
        else if(strcmp(argv[idx], "--name-only") == 0)
        {
            g_name_only = true;
        }
        else if(strcmp(argv[idx], "-V") == 0
             || strcmp(argv[idx], "--version") == 0)
        {
            std::cout << LIBADDR_VERSION_STRING << '\n';
            return 0;
        }
        else
        {
            std::cerr
                << "error: unknown command line option \""
                << argv[idx]
                << "\". Try --help for additional info.\n";
            return 1;
        }
    }

    addr::iface::pointer_vector_t interfaces(addr::iface::get_local_addresses());
    if(interfaces->empty())
    {
        std::cerr << "error: no interfaces found, is your network up?" << std::endl;
        return 1;
    }

    // headers
    //
    if(!g_hide_headers && !g_name_only)
    {
        std::cout << "Iface            "
                     "Flags "
                     "Address                                  "
                     "Broadcast                                "
                     "Destination                              "
                     "\n";
    }

    std::string default_interface;
    if(g_filter == FILTER_DEFAULT)
    {
        addr::route::vector_t routes(addr::route::get_ipv4_routes());
        if(routes.empty())
        {
            std::cerr << "error: unknown default route.\n";
            return 1;
        }
        for(auto const & r : routes)
        {
            if(r->get_destination_address().is_default())
            {
                default_interface = r->get_interface_name();
            }
        }
    }

    std::set<std::string> found;
    for(auto const & i : *interfaces)
    {
        addr::addr const & a(i->get_address());
        switch(g_filter)
        {
        case FILTER_ALL:
            // no filtering
            break;

        case FILTER_PUBLIC:
            if(a.get_network_type() != addr::network_type_t::NETWORK_TYPE_PUBLIC)
            {
                continue;
            }
            break;

        case FILTER_PRIVATE:
            if(a.get_network_type() != addr::network_type_t::NETWORK_TYPE_PRIVATE)
            {
                continue;
            }
            break;

        case FILTER_LOOPBACK:
            if(a.get_network_type() != addr::network_type_t::NETWORK_TYPE_LOOPBACK)
            {
                continue;
            }
            break;

        case FILTER_DEFAULT:
            if(a.get_name() != default_interface)
            {
                continue;
            }
            break;

        }

        if(g_name_only)
        {
            // prevent printing the same name twice
            //
            std::pair inserted(found.insert(i->get_name()));
            if(inserted.second)
            {
                std::cout
                    << i->get_name()
                    << '\n';
            }
        }
        else
        {
            addr::addr const & b(i->get_broadcast_address());
            addr::addr const & d(i->get_destination_address());
            bool const a_is_ipv4(a.is_ipv4());

            addr::string_ip_t const mode(addr::STRING_IP_ADDRESS
                | (g_asterisk ? addr::STRING_IP_DEFAULT_AS_ASTERISK : 0)
                | (a_is_ipv4 ? addr::STRING_IP_DEFAULT_AS_IPV4 : 0));

            std::cout
                << std::left << std::setw(17) << i->get_name()
                << std::left << std::setw( 6) << i->get_flags()  // TODO: add conversion to letters
                << std::left << std::setw(41) << a.to_ipv4or6_string(addr::STRING_IP_ADDRESS)
                << std::left << std::setw(41) << b.to_ipv4or6_string(mode)
                << std::left << std::setw(41) << d.to_ipv4or6_string(mode)
                << '\n';
        }
    }

    return 0;
}

// vim: ts=4 sw=4 et

